src/System.Runtime.Extensions/src/System/AppDomain.cs

```diff
@@ -280,5 +282,45 @@ public void SetThreadPrincipal(IPrincipal principal)
                 _defaultPrincipal = principal;
             }
         }
+
+        public IPrincipal GetThreadPrincipal()
+        {
+            IPrincipal principal = null;
+            if (_defaultPrincipal != null)
+            {
+                principal = _defaultPrincipal;
+            }
+            else
+            {
+                switch (_principalPolicy)
+                {
+                    case PrincipalPolicy.NoPrincipal:
+                        principal = null;
+                        break;
+                    case PrincipalPolicy.UnauthenticatedPrincipal:
+                        Assembly assembly = Assembly.Load(new AssemblyName("System.Security.Claims"));
+                        Type genericPrincipal = assembly?.GetType("System.Security.Principal.GenericPrincipal");
+                        Type genericIdentity = assembly?.GetType("System.Security.Principal.GenericIdentity");
+
+                        ConstructorInfo genericPrincipalCtor = genericPrincipal?.GetConstructor( new[] { genericIdentity, typeof(string[])});
+                        ConstructorInfo genericIdentityCtor = genericIdentity?.GetConstructor( new[] { typeof(string), typeof(string) });
+
+                        principal = (IPrincipal)genericPrincipalCtor.Invoke(new object[] { genericIdentityCtor.Invoke( new object[] { "", "" }), new string[] { "" } });
+                        break;
+                    case PrincipalPolicy.WindowsPrincipal:
+                        assembly = Assembly.Load(new AssemblyName("System.Security.Principal.Windows"));
+                        Type windowsPrincipal = assembly?.GetType("System.Security.Principal.WindowsPrincipal");
+                        Type windowsIdentity = assembly?.GetType("System.Security.Principal.WindowsIdentity");
+
+                        ConstructorInfo windowsPrincipalCtor = windowsPrincipal?.GetConstructor( new[] { windowsIdentity });
+                        MethodInfo currentIdentity = windowsIdentity?.GetMethod("GetCurrent", new Type[] { } );
+
+                        principal = (IPrincipal)windowsPrincipalCtor?.Invoke( new object[] { currentIdentity.Invoke(null, new object[] { } ) });
+                        break;
```

Comment:
There's a lot of overhead being generated here using reflection to get the relevant objects every time GetThreadPrincipal is called.  Would it make sense to cache the relevant objects on first use such that on subsequent use we don't have to rediscover all of these objects?